lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -4724,6 +4728,72 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+FIRRTLType RWProbeOp::inferReturnType(ValueRange operands,
+                                      ArrayRef<NamedAttribute> attrs,
+                                      std::optional<Location> loc) {
+  auto typeAttr = getAttr<TypeAttr>(attrs, "type");
+  auto type = typeAttr.getValue();
+  auto baseType = type_dyn_cast<FIRRTLBaseType>(type);
+  if (!baseType)
+    return emitInferRetTypeError(loc, "must be base type, not ", type);
+  if (baseType.hasUninferredWidth() || baseType.hasUninferredReset())
+    return emitInferRetTypeError(
+        loc, "must not have uninferred width or reset in type ", baseType);
+  auto forceableType = firrtl::detail::getForceableResultType(true, baseType);
+  if (!forceableType)
+    return emitInferRetTypeError(loc, "cannot force type ", type);
+  return forceableType;
+}
+
+LogicalResult RWProbeOp::verifyInnerRefs(hw::InnerRefNamespace &ns) {
+  auto targetRef = getTarget();
+  if (!targetRef)
+    return emitOpError("has invalid target reference");
+  if (targetRef.getModule() !=
+      (*this)->getParentOfType<FModuleLike>().getModuleNameAttr())
+    return emitOpError() << "has non-local target";
+
+  auto target = ns.lookup(targetRef);
+  if (!target)
+    return emitOpError() << "has target that cannot be resolved: " << target;
+
+  auto getFinalType = [&](auto type, auto fieldID,
+                          Location loc) -> FailureOr<Type> {
+    auto fieldIDType = type_dyn_cast<hw::FieldIDTypeInterface>(type);
+    if (fieldIDType)
+      return fieldIDType.getFinalTypeByFieldID(fieldID);
+    // Verifier on target op should handle this, check to be safe anyway.
+    if (fieldID != 0) {
+      auto diag =
+          emitOpError()
+          << "has target that resolves to invalid per-field inner symbol";
+      return diag.attachNote(loc)
+             << "target does not support per-field inner symbols";
+    }
+    return type;
+  };
+  auto checkFinalType = [&](auto type, Location loc) -> LogicalResult {
+    auto fType = getFinalType(type, target.getField(), loc);
+    if (failed(fType))
+      return failure();
+    if (*fType != getType())
+      return emitOpError("has type mismatch: target resolves to ")
+             << *fType << " instead of expected " << getType();
+    return success();
+  };
+  if (target.isPort()) {
+    auto mod = cast<FModuleLike>(target.getOp());
+    return checkFinalType(mod.getPortType(target.getPort()),
+                          mod.getPortLocation(target.getPort()));
+  } else {
+    hw::InnerSymbolOpInterface symOp =
+        cast<hw::InnerSymbolOpInterface>(target.getOp());
+    if (!symOp.getTargetResult())
+      return emitOpError("has target that does not resolve to a result");
```

Comment:
Suggestion: This error message may be too low-level. Something more like: "targets an operation which does not have exactly one result".

Reply:
Does `getTargetResult()` do the right thing if this is an instance that has a single port?

Reply:
> Does `getTargetResult()` do the right thing if this is an instance that has a single port?

Yes, that's the reason it exists actually (so I hope so)-- to avoid encoding per-op semantics of inner symbols (fixing bugs re:InstanceOp was part of motivation).  InstanceOp should report:

* no target result -- inner symbol never targets/names a result, only the op itself
* no support for per-field inner symbols -- (mostly result of the above-- default impl says this isn't supported if no target result is named in particular)

I'll see about tweaking the language here, agreed it's wrong level.